### PR TITLE
Better subscribe mechanism

### PR DESCRIPTION
**Pull request by kelunik**
Will be `v0.3.0`.

Opinions? @amphp/database @vuliad @glowdan

### CODE REVIEW

File: lib/SubscribeClient.php [hunk 1]
```diff
@@ -166,85 +147,91 @@ private function applyUri($uri) {
...
-     * @return Promise
+     * @return Subscription
      */
     public function pSubscribe($pattern) {
         $promisor = new Deferred;
+        $subscription = new Subscription($promisor, function($promisor) use ($pattern) {
+            $this->unloadPatternPromisor($promisor, $pattern);
+        });
+
+        $this->patternPromisors[$pattern][spl_object_hash($promisor)] = $promisor;
```
Comment (bwoebi) [inline lib/SubscribeClient.php]: Why don't you just use() the $promisor too and just pass the promise (and not the Promisor) to Subscription?

File: lib/Subscription.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        return $this;
+    }
+
+    public function cancel() {
+        $cb = $this->unsubscribeCallback;
+        $cb($this->promisor);
+    }
+
+    public function __destruct() {
+        $this->cancel();
```
Comment (bwoebi) [inline lib/Subscription.php]: Are you sure that's a great idea? You may want to hold a Subscription open indefinitely (i.e. just set the watch()/when() and then forget about the object?

**Comment by bwoebi** [comment]
In general this looks like a more stable approach, looks great :-)

File: lib/SubscribeClient.php (hunk 1 above)
Comment (kelunik) [inline lib/SubscribeClient.php]: Good point.

File: lib/Subscription.php (hunk 2 above)
Comment (kelunik) [inline lib/Subscription.php]: Removed it. I thought you usually keep the object around, but that's probably not the case for most use cases.

File: lib/Subscription.php [hunk 3]
```diff
@@ -10,28 +10,28 @@ class Subscription implements Promise {
     /** @var callable */
     private $unsubscribeCallback;
     /** @var Deferred */
```
Comment (bwoebi) [inline lib/Subscription.php]: Bad typehint … ;-)

File: lib/Subscription.php [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+namespace Amp\Redis;
+
+use Amp\Deferred;
+use Amp\Promise;
+use Amp\Promisor;
+
+class Subscription implements Promise {
+    /** @var callable */
+    private $unsubscribeCallback;
+    /** @var Deferred */
```
Comment (bwoebi) [inline lib/Subscription.php]: bad typehint … ;-)

Comment (kelunik) [inline lib/Subscription.php]: PhpStorm didn't complain.

Comment (bwoebi) [inline lib/Subscription.php]: But isn't the typehint anyway implied in phpstorm redundant due to assign in ctor?

Comment (kelunik) [inline lib/Subscription.php]: It wasn't, don't know why.

**Pull request closed without merging**